### ISSUE TITLE
[java] UnusedPrivateMethod: false positive caused by mismatch between method visibility

### ISSUE THREAD

**Issue by punkratz312** [pmd/pmd#5271]
This legacy code is written in an imperative style, coupled through fields instead of using a fluid API. I recommend merging the cleanup, as it's primarily a refactoring effort to locate the actual bug, with the tests still passing.

https://github.com/pmd/pmd/pull/5268

@jsotuyod please see yourself:

failures:
```
public method called by private method
private method called by private method
public method not called by public method
public and public method not called
public and private method not called
```
tests:
```
<?xml version="1.0" encoding="UTF-8"?>
<test-data
        xmlns="http://pmd.sourceforge.net/rule-tests"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="http://pmd.sourceforge.net/rule-tests http://pmd.sourceforge.net/rule-tests_1_0_0.xsd">
    <test-code>
        <description>public method called by public method</description>
        <expected-problems>0</expected-problems>
        <code><![CDATA[
                public class Foo {
                    public void bar() {
                        fooPublic();
                    }
                    public void fooPublic() {}
                }
            ]]>
        </code>
    </test-code>
    <test-code>
        <description>public method called by private method</description>
        <expected-problems>0</expected-problems>
        <code><![CDATA[
                public class Foo {
                    private void bar() {
                        fooPublic();
                    }
                    public void fooPublic() {}
                }
            ]]>
        </code>
    </test-code>
    <test-code>
        <description>private method called by public method</description>
        <expected-problems>0</expected-problems>
        <code><![CDATA[
                public class Foo {
                    public void bar() {
                        fooPrivate();
                    }
                    private void fooPrivate() {}
                }
            ]]>
        </code>
    </test-code>
    <test-code>
        <description>private method called by private method</description>
        <expected-problems>0</expected-problems>
        <code><![CDATA[
                public class Foo {
                    private void bar() {
                        fooPrivate();
                    }
                    private void fooPrivate() {}
                }
            ]]>
        </code>
    </test-code>
    <test-code>
        <description>public method not called by public method</description>
        <expected-problems>1</expected-problems>
        <code><![CDATA[
                public class Foo {
                    public void bar() {
                    }
                    public void fooPublic() {}
                }
            ]]>
        </code>
    </test-code>
    <test-code>
        <description>private method not called by public method</description>
        <expected-problems>1</expected-problems>
        <code><![CDATA[
                public class Foo {
                    public void bar() {
                    }
                    private void fooPrivate() {}
                }
            ]]>
        </code>
    </test-code>
    <test-code>
        <description>public and public method not called</description>
        <expected-problems>2</expected-problems>
        <code><![CDATA[
                public class Foo {
                    public void bar() {}
                    public void fooPublic() {}
                }
            ]]>
        </code>
    </test-code>
    <test-code>
        <description>public and private method not called</description>
        <expected-problems>2</expected-problems>
        <code><![CDATA[
                public class Foo {
                    public void bar() {}
                    private void fooPrivate() {}
                }
            ]]>
        </code>
    </test-code>
    <test-code>
        <description>private and private method not called</description>
        <expected-problems>2</expected-problems>
        <code><![CDATA[
                public class Foo {
                    private void bar() {}
                    private void fooPrivate() {}
                }
            ]]>
        </code>
    </test-code>
```
<img width="1785" alt="image" src="https://github.com/user-attachments/assets/43edd44b-424a-4683-8b89-1f650b009cdd">

**Comment by jsotuyod** [comment]
The rule is called `Unused**Private**Method`, it will only flag private methods, as it's not possible to know which public methods are in use (it would require multifile analysis which is not supported, and even then it may not be enough, ie: a library will expose lots of public methods that are never called from the main sourceset).

So, of the provided examples, most are automatically outside the scope of the rule.

For others, such as `private method called by private method`, the test itself is wrong. You have 2 private methods, 1 of them is called from the other, but the second one is never used, so it's flagged, and the test will find 1 violation, not 0.

The rule seems to be performing as expected on these scenarios.

**Comment by punkratz312** [comment]
Sorry, I missed that. We only need the unused method, but it makes sense now. Thanks for the clarification!
